transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -0,0 +1,2272 @@
+//! This plays a similar role to the transaction builder.
+
+use crate::{ChangeDestination, InputCredentials, MemoBuilder, TxBuilderError};
+use core::cmp::min;
+use mc_account_keys::PublicAddress;
+use mc_fog_report_validation::FogPubkeyResolver;
+use mc_transaction_core::{
+    ring_signature::{GeneratorCache, OutputSecret, RingMLSAG, Scalar, SignableInputRing},
+    tx::{TxIn, TxOut, TxOutConfirmationNumber},
+    Amount, BlockVersion, InputRules, MemoContext, MemoPayload, NewMemoError,
+    SignedContingentInput, UnmaskedAmount,
+};
+use rand_core::{CryptoRng, RngCore};
+
+/// Helper utility for creating signed contingent inputs with required outputs,
+/// and attaching fog hint and memos as appropriate.
+///
+/// This is generic over FogPubkeyResolver because there are several reasonable
+/// implementations of that.
+///
+/// This is generic over MemoBuilder to allow injecting a policy for how to
+/// use the memos in the TxOuts.
+#[derive(Debug)]
+pub struct SignedContingentInputBuilder<FPR: FogPubkeyResolver> {
+    /// The block version that we are targeting for this transaction
+    block_version: BlockVersion,
+    /// The input material used to form the transaction
+    input_credentials: InputCredentials,
+    /// Global indices for the tx out's in the ring
+    tx_out_global_indices: Vec<u64>,
+    /// The outputs created by the transaction, and associated secrets
+    outputs_and_secrets: Vec<(TxOut, OutputSecret)>,
+    /// The tombstone_block value, a block index in which the transaction
+    /// expires, and can no longer be added to the blockchain
+    tombstone_block: u64,
+    /// The source of validated fog pubkeys used for this transaction
+    fog_resolver: FPR,
+    /// The limit on the tombstone block value imposed pubkey_expiry values in
+    /// fog pubkeys used so far
+    fog_tombstone_block_limit: u64,
+    /// An policy object implementing MemoBuilder which constructs memos for
+    /// this transaction.
+    ///
+    /// This is an Option in order to allow working around the borrow checker.
+    /// Box<dyn ...> is used because having more generic parameters creates more
+    /// types that SDKs must bind to if they support multiple memo builder
+    /// types.
+    memo_builder: Option<Box<dyn MemoBuilder + 'static + Send + Sync>>,
+}
+
+impl<FPR: FogPubkeyResolver> SignedContingentInputBuilder<FPR> {
+    /// Initializes a new SignedContingentInputBuilder.
+    ///
+    /// # Arguments
+    /// * `block_version` - The block version rules to use when signing the
+    ///   input
+    /// * `input_credentials` - Credentials for the input we are signing
+    /// * `tx_out_global_indices` - Global indices for the tx out's in the ring
+    /// * `fog_resolver` - Source of validated fog keys to use with this
+    ///   transaction
+    /// * `memo_builder` - An object which creates memos for the TxOuts in this
+    ///   transaction
+    pub fn new<MB: MemoBuilder + 'static + Send + Sync>(
```

Comment:
OOC, why is `'static` required here but not in `new_with_box`?

Reply:
I believe you need `'static` here so that you could `Box::new` the object. What I think `'static` implies when used as a trait bound is that the object does not contain any fields with references that have non-`'static ` lifetimes, or to put it differently - it is safe to move the object around knowing that it does not point at anything that could potentially go away.
From reading https://doc.rust-lang.org/reference/lifetime-elision.html#default-trait-object-lifetimes I am  getting that `'static` is implied in a `Box<dyn ...>` so it is not explicitly required to be specified, due to `dyn` implying `'static` when an explicit lifetime is not specified.